# Indexing

#### Setting up the data

Let's create the structures that will be used later in this notebook

In [1]:
import numpy as np

In [2]:
np.random.seed(42)  # Setting the random seed

In [3]:
# a vector: the argument to the array function is a Python list
v = np.random.rand(10)
v

array([0.37454012, 0.95071431, 0.73199394, 0.59865848, 0.15601864,
       0.15599452, 0.05808361, 0.86617615, 0.60111501, 0.70807258])

In [4]:
# a matrix: the argument to the array function is a nested Python list
M = np.random.rand(10, 2)
M

array([[0.02058449, 0.96990985],
       [0.83244264, 0.21233911],
       [0.18182497, 0.18340451],
       [0.30424224, 0.52475643],
       [0.43194502, 0.29122914],
       [0.61185289, 0.13949386],
       [0.29214465, 0.36636184],
       [0.45606998, 0.78517596],
       [0.19967378, 0.51423444],
       [0.59241457, 0.04645041]])

We can index elements in an array using the square bracket and indices:

In [5]:
# v is a vector, and has only one dimension, taking one index
v[0]

0.3745401188473625

In [6]:
# M is a matrix, or a 2 dimensional array, taking two indices 
M[1,1]

0.21233911067827616

If we omit an index of a multidimensional array it returns the whole row (or, in general, a N-1 dimensional array) 

In [7]:
M[1] 

array([0.83244264, 0.21233911])

The same thing can be achieved with using `:` instead of an index: 

In [8]:
M[1,:] # row 1

array([0.83244264, 0.21233911])

In [9]:
M[:,1] # column 1

array([0.96990985, 0.21233911, 0.18340451, 0.52475643, 0.29122914,
       0.13949386, 0.36636184, 0.78517596, 0.51423444, 0.04645041])

We can assign new values to elements in an array using indexing:

In [10]:
M[0,0] = 1

In [11]:
M

array([[1.        , 0.96990985],
       [0.83244264, 0.21233911],
       [0.18182497, 0.18340451],
       [0.30424224, 0.52475643],
       [0.43194502, 0.29122914],
       [0.61185289, 0.13949386],
       [0.29214465, 0.36636184],
       [0.45606998, 0.78517596],
       [0.19967378, 0.51423444],
       [0.59241457, 0.04645041]])

In [12]:
# also works for rows and columns
M[1,:] = 0
M[:,1] = -1

In [13]:
M

array([[ 1.        , -1.        ],
       [ 0.        , -1.        ],
       [ 0.18182497, -1.        ],
       [ 0.30424224, -1.        ],
       [ 0.43194502, -1.        ],
       [ 0.61185289, -1.        ],
       [ 0.29214465, -1.        ],
       [ 0.45606998, -1.        ],
       [ 0.19967378, -1.        ],
       [ 0.59241457, -1.        ]])

## Index slicing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array:

In [14]:
a = np.array([1,2,3,4,5])
a

array([1, 2, 3, 4, 5])

In [15]:
a[1:3]

array([2, 3])

Array slices are **mutable**: if they are assigned a new value the original array from which the slice was extracted is modified:

In [16]:
a[1:3] = [-2,-3]

a

array([ 1, -2, -3,  4,  5])

* We can omit any of the three parameters in `M[lower:upper:step]`:

In [17]:
a[::] # lower, upper, step all take the default values

array([ 1, -2, -3,  4,  5])

In [18]:
a[::2] # step is 2, lower and upper defaults to the beginning and end of the array

array([ 1, -3,  5])

In [19]:
a[:3] # first three elements

array([ 1, -2, -3])

In [20]:
a[3:] # elements from index 3

array([4, 5])

* Negative indices counts from the end of the array (positive index from the begining):

In [21]:
a = np.array([1,2,3,4,5])

In [22]:
a[-1] # the last element in the array

5

In [23]:
a[-3:] # the last three elements

array([3, 4, 5])

* Index slicing works exactly the same way for multidimensional arrays:

In [24]:
A = np.array([[n+m*10 for n in range(5)] 
              for m in range(5)])
A

array([[ 0,  1,  2,  3,  4],
       [10, 11, 12, 13, 14],
       [20, 21, 22, 23, 24],
       [30, 31, 32, 33, 34],
       [40, 41, 42, 43, 44]])

In [25]:
# a block from the original array
A[1:4, 1:4]

array([[11, 12, 13],
       [21, 22, 23],
       [31, 32, 33]])

In [26]:
# strides
A[::2, ::2]

array([[ 0,  2,  4],
       [20, 22, 24],
       [40, 42, 44]])

### Indexing and Array Memory Management

Numpy arrays support two different way of storing data into memory, namely

* F-Contiguous 
    - i.e. *column-wise* storage, Fortran-like
* C-Contiguous
    - i.e. *row-wise* storage, C-like
    
The **storage** strategy is controlled by the parameter `order` of `np.array`

<img src="images/storage_simple.png" />

## Fancy indexing

In [27]:
import numpy as np
FC = np.array([[1, 2, 3],    [4, 5, 6], 
               [7, 8, 9],    [10, 11, 12]], order='F')

In [28]:
CC = np.array([[1, 2, 3],    [4, 5, 6], 
               [7, 8, 9],    [10, 11, 12]], order='C')

* **Note**: no changes in meaning for indexing operations

In [29]:
FC[0, 1]

2

In [30]:
CC[0, 1]

2

In [31]:
FC.shape

(4, 3)

In [32]:
CC.shape

(4, 3)

<img src="images/storage_index.png" />

## Fancy Indexing

Fancy indexing is the name for when an array or list is used in-place of an index: 

In [33]:
row_indices = [1, 2, 3]
A[row_indices]

array([[10, 11, 12, 13, 14],
       [20, 21, 22, 23, 24],
       [30, 31, 32, 33, 34]])

In [34]:
col_indices = [1, 2, -1] # remember, index -1 means the last element
A[row_indices, col_indices]

array([11, 22, 34])

* We can also index **masks**: 

    - If the index mask is an Numpy array of with data type `bool`, then an element is selected (True) or not (False) depending on the value of the index mask at the position each element: 

In [35]:
b = np.array([n for n in range(5)])
b

array([0, 1, 2, 3, 4])

In [36]:
row_mask = np.array([True, False, True, False, False])
b[row_mask]

array([0, 2])

* Alternatively:

In [37]:
# same thing
row_mask = np.array([1,0,1,0,0], dtype=bool)
b[row_mask]

array([0, 2])

This feature is very useful to conditionally select elements from an array, using for example comparison operators:

In [38]:
x = np.arange(0, 10, 0.5)
x

array([0. , 0.5, 1. , 1.5, 2. , 2.5, 3. , 3.5, 4. , 4.5, 5. , 5.5, 6. ,
       6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

In [39]:
mask = (5 < x)

mask

array([False, False, False, False, False, False, False, False, False,
       False, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True])

In [40]:
x[mask]

array([5.5, 6. , 6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

Alternatively, we can use the condition (mask) array directly within brackets to index the array

In [41]:
x[(5 < x)]

array([5.5, 6. , 6.5, 7. , 7.5, 8. , 8.5, 9. , 9.5])

---

# Exercises on Indexing

Index slicing is the technical name for the syntax `M[lower:upper:step]` to extract part of an array

## Ex 3.1 

Generate a three-dimensional array of any size containing random numbers taken from an uniform distribution (_guess the numpy function in `np.random`_). Then print out separately the first entry along the three axis (i.e. `x, y, z`)  


* _hint_: Slicing with numpy arrays works quite like Python lists

## Ex 3.2

Create a vector and print out elements in reverse order

#### Hint: Use slicing for this exercise

## Ex 3.3

Generate a $7 \times 7$ matrix and replace all the elements in odd rows and even columns with `1`.

#### Hint: Use slicing to solve this exercise!

#### Note: Take a look at the original matrix, then.

**Use fancy indexing** to get all the elements of the previous matrix that are equals to `1`

## Ex 3.4 

Generate a `10 x 10` matrix of numbers `A`. Then, generate a numpy array of integers in range `1-9`. Pick `5` random values (with no repetition) from this array and use these values to extract rows from the original matrix `A`.

## Ex 3.5 

Repeat the previous exercise but this time extract columns from `A`

## Ex 3.6

Generate an array of numbers from `0` to `20` with step `0.5`. 
Extract all the values greater than a randomly generated number in the same range.

#### Hint: Try to write the condition as an expression and save it to a variable. Then, use this variable in square brackets to index.... this is when the magic happens!